cachito/workers/tasks/gomod.py

```diff
@@ -16,45 +17,84 @@
+            )
+
+
 @app.task
-def fetch_gomod_source(request_id, dep_replacements=None):
+def fetch_gomod_source(request_id, dep_replacements=None, package_configs=None):
     """
     Resolve and fetch gomod dependencies for a given request.
 
     :param int request_id: the Cachito request ID this is for
     :param list dep_replacements: dependency replacements with the keys "name" and "version"
+    :param list package_configs: the list of optional package configurations submitted by the user
     :raises CachitoError: if the dependencies could not be retrieved
     """
     config = get_worker_config()
+    if package_configs is None:
+        package_configs = []
+
     bundle_dir = RequestBundleDir(request_id)
+    log.debug("Checking if the application source uses golang")
+    subpaths = [os.path.normpath(c["path"]) for c in package_configs if c.get("path")]
+
+    if not subpaths:
+        # Default to the root of the application source
+        subpaths = [os.curdir]
+
+    _verify_gomod_files(bundle_dir, subpaths, config.cachito_gomod_ignore_missing_gomod_file)
 
-    if not bundle_dir.go_mod_file.exists():
-        if config.cachito_gomod_ignore_missing_gomod_file:
-            log.warning("The go.mod file is missing for request %d", request_id)
-            return
-
-        raise CachitoError("The go.mod file is missing")
-
-    log.info("Fetching gomod dependencies for request %d", request_id)
-    request = set_request_state(request_id, "in_progress", "Fetching the gomod dependencies")
-    try:
-        gomod = resolve_gomod(str(bundle_dir.source_dir), request, dep_replacements)
-    except CachitoError:
-        log.exception("Failed to fetch gomod dependencies for request %d", request_id)
-        raise
-
-    env_vars = {
-        "GOCACHE": {"value": "deps/gomod", "kind": "path"},
-        "GOPATH": {"value": "deps/gomod", "kind": "path"},
-    }
-    env_vars.update(config.cachito_default_environment_variables.get("gomod", {}))
-    update_request_with_package(request_id, gomod["module"], env_vars)
-    update_request_with_deps(request_id, gomod["module"], gomod["module_deps"])
-
-    # add package deps
-    for package in gomod["packages"]:
-        if package.get("pkg_deps"):
-            # This also adds the package to the request
-            update_request_with_deps(request_id, package["pkg"], package["pkg_deps"])
+    for i, subpath in enumerate(subpaths):
+        log.info(
+            "Fetching the gomod dependencies for request %d in subpath %s", request_id, subpath
+        )
+        request = set_request_state(
+            request_id,
+            "in_progress",
+            f'Fetching the gomod dependencies at the "{subpath}" directory',
+        )
+        gomod_source_path = str(bundle_dir.app_subpath(subpath).source_dir)
+        try:
+            gomod = resolve_gomod(gomod_source_path, request, dep_replacements)
```

Comment:
@athos-ribeiro, in the npm tasks, it looks like there is some efficiency introduced here to prevent resolving dependencies

```
            package_and_deps_info = resolve_npm(
                package_source_path, request, skip_deps=downloaded_deps
            )
```

We could potentially do something like that here, but I did not chase that down since I was not sure since it seemed like premature optimization.

Reply:
Agreed, I wouldn't worry about this unless it becomes an issue